rust/src/lib.rs

```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
+/// # Safety
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must
+///   have an alignment of `1`. Its contents must be a string representing a valid system
+///   path in the operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the
+///   function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `proposal_ptr` must be non-null and valid for reads for `proposal_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `proposal_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `proposal_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `ufvk` must be non-null and must point to a null-terminated UTF-8 string.
+/// - Call [`zcashlc_free_boxed_slice`] to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_create_pczt_from_proposal(
+    db_data: *const u8,
+    db_data_len: usize,
+    network_id: u32,
+    proposal_ptr: *const u8,
+    proposal_len: usize,
+    ufvk: *const c_char,
+) -> *mut FfiBoxedSlice {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let mut db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
+
+        let proposal =
+            Proposal::decode(unsafe { slice::from_raw_parts(proposal_ptr, proposal_len) })
+                .map_err(|e| anyhow!("Invalid proposal: {}", e))?
+                .try_into_standard_proposal(&db_data)?;
+
+        let ufvk_str = unsafe { CStr::from_ptr(ufvk).to_str()? };
+        let ufvk = UnifiedFullViewingKey::decode(&network, ufvk_str).map_err(|e| {
+            anyhow!(
+                "Value \"{}\" did not decode as a valid UFVK: {}",
+                ufvk_str,
+                e
+            )
+        })?;
+
+        let account_id = db_data
+            .get_account_for_ufvk(&ufvk)
+            .map_err(|e| anyhow!("Error retrieving account information for ufvk: {}", e))?
+            .ok_or(anyhow!(
+                "UFVK does not correspond to an account in the wallet."
+            ))?
+            .id();
+
+        if proposal.steps().len() == 1 {
+            let pczt = create_pczt_from_proposal::<_, _, Infallible, _, Infallible, _>(
+                &mut db_data,
+                &network,
+                account_id,
+                OvkPolicy::Sender,
+                &proposal,
+            )
+            .map_err(|e| anyhow!("Error creating PCZT from single-step proposal: {}", e))?;
+
+            Ok(FfiBoxedSlice::some(pczt.serialize()))
+        } else {
+            Err(anyhow!(
+                "Multi-step proposals are not yet supported for PCZT generation."
```

Comment:
Should we actually be returning an array of PCZTs?

Reply:
For a multi-step proposal, we will likely want a different API; there will be some cases where we can construct all of the PCZTs for the steps at once, but that's not the case for every multi-step proposal (for example, if the first step requires a multisig PCZT, I don't think we can necessarily construct the PCZT for the second step.)